torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
+  // every step. We're just missing a few functions for ATen, but the fix
+  // should be straightforward.
+  int64_t dim_after_broadcast = 0;
+  for (int64_t arg_idx : arg_subset) {
+    dim_after_broadcast = std::max(dim_after_broadcast, args[arg_idx].dim());
+  }
+  std::vector<int64_t> map_size;
+  for (size_t i = 0; i < arg_subset.size(); ++i) {
+    auto & arg = args.at(arg_subset[i]);
+    auto & chunk_desc = input_chunks.at(arg_subset[i]);
+    if (chunk_desc.nSubtensors == 1) {
+      try {
+        map_size = at::infer_size(map_size, arg.sizes());
+      } catch (std::exception& e) {
+        return at::nullopt;
+      }
     } else {
-      JIT_ASSERT(canCompileOnCPU());
-      raw_func = new CPUFusionFunction(name, agraph, config_);
+      auto tensor_sizes = arg.sizes().vec();
+      int64_t num_chunks = chunk_desc.nSubtensors;
+      int64_t dim = chunk_desc.dim;
+      if (dim < 0) {
+        dim += arg.dim();
+      }
+      JIT_ASSERT(dim >= 0 && dim <= static_cast<int64_t>(tensor_sizes.size()));
+      if (tensor_sizes[dim] % num_chunks != 0) {
+        return at::nullopt;
+      }
+      tensor_sizes[dim] /= num_chunks;
+      try {
+        map_size = at::infer_size(map_size, tensor_sizes);
+      } catch (std::exception& e) {
+        return at::nullopt;
+      }
     }
-    it = cache.emplace(key_, std::shared_ptr<CompiledFusionFunction>(raw_func)).first;
   }
-  return it->second;
+
+  return {map_size};
 }
 
-std::shared_ptr<CompiledFusionFunction> FusionCompiler::getOrCompile(Node* fusion_group) {
-  auto & graph = *fusion_group->g(attr::Subgraph);
-  AnnotatedGraph agraph(graph, fusion_group->i(attr::device));
-  for(auto & input : graph.inputs()) {
-    auto t = input->type()->expect<TensorType>();
-    agraph.input_desc.emplace_back(t);
+// See Note [Run-time shape checking code] for more explanation on the algorithm.
+at::optional<std::vector<int64_t>> FusedKernelCache::canRunKernel(at::TensorList args) {
+  AT_CHECK(args.size() == input_chunks.size(),
+           "Expected ", input_chunks.size(), " arguments, but got ", args.size());
+
+  at::optional<std::vector<int64_t>> map_size;
+  for (const auto & broadcast_group : input_broadcast_groups) {
+    if (!map_size) {
+      map_size = getMapSize(args, broadcast_group);
+      if (!map_size) {
+        return at::nullopt;
+      }
+    } else {
+      auto group_map_size = getMapSize(args, broadcast_group);
+      // NB: this checks that group_map_size is defined AND equal to map_size
+      if (map_size != group_map_size) {
+        return at::nullopt;
+      }
+    }
   }
-  for(auto & output : graph.outputs()) {
-    auto t = output->type()->expect<TensorType>();
-    agraph.output_desc.emplace_back(t);
+  return map_size;
+}
+
+void FusedKernelCache::runFallback(Stack& stack) {
+  InterpreterState(fallback_code).runOneStage(stack);
+}
+
+void FusedKernelCache::expandArgs(std::vector<at::Tensor>& args, std::vector<int64_t>& map_size) {
```

Comment:
Ugh, pass by mutable reference :(

Reply:
well I only mutate it's elements, and doing `(*args)[i]` is so awful, that I doubt the "mutable reference" rule wins here over "make the code readable".

Reply:
If I didn't want to mutate it, it would get passed as a `const` reference, or as an `ArrayRef`